folding/utils/runandlog.py

```diff
@@ -0,0 +1,142 @@
+import subprocess
+import tqdm
+from typing import List, Dict
+from collections import defaultdict
+import bittensor as bt 
+import copy
+import wandb
+from loguru import logger
+from dataclasses import asdict, dataclass
+
+import folding
+
+
+"""
+We need something thet gets returned (result dict) 
+We don't want it to grow each time its called
+We want it to be independent of gromacs
+We dont need to be worried about pdb_id 
+This needs to log to the event dict 
+"""
+
+
+def output_dict():
+    return defaultdict(str)
+
+class CommandFailedException(Exception):
+    pass
+
+class RunAndLog():
+    def __init__(self):
+        self.command_dict = defaultdict(output_dict)
+    
+    def run_cmd_commands(
+            self, commands: List [str], suppress_cmd_output: bool = True, verbose: bool = False
+    ) -> Dict[str, str]:
+        
+        for cmd in tqdm.tqdm(commands): # set tuple 
+            bt.logging.debug(F"Running command {cmd}")
+
+            try:
+                # result = run the command (cmd), check that it succedded, executed it through the shell, captures its output and error messages. 
+                result = subprocess.run( 
+                    cmd, 
+                    check=True, 
+                    shell=True, 
+                    stdout=subprocess.PIPE, 
+                    stderr=subprocess.PIPE, 
+                )
+                if not suppress_cmd_output:
+                    bt.logging.info(result.stdout.decode())
+
+                    # no need to record and log successful command executions
+
+            except subprocess.CalledProcessError as e:                
+                # this will always print to the terminal regardless if verbose.
+                bt.logging.error(f"❌ Failed to run command ❌: {cmd}") 
+
+                if verbose:
+                    bt.logging.error(f"Output: {e.stdout.decode()}")
+                    bt.logging.error(f"Error: {e.stderr.decode()}")
+
+                # update command_dict
+                self.command_dict[str(cmd)]["status"] = 'failed'
```

Comment:
if there is an error, it is implicit that the command failed. You dont need this

Reply:
maybe it would be also nice to not have the entire command, but also just `self.command_dict[str(cmd.split('gmx').split()[0])]` so you just grab the name of the command right after the gmx string? Not necessary. We could also do this during analysis

Reply:
great idea, no need for the other args

Reply:
https://github.com/macrocosm-os/folding/pull/147/commits/4533a6b25d6c6dbe0f5d0066bb176b6cfa871e04